### ISSUE TITLE
Cap fails to start when SCM plugin is installed too early in Capfile

### ISSUE THREAD

**Issue by mattbrictson** [capistrano/capistrano#1824]
Originally reported on Stack Overflow here: http://stackoverflow.com/q/41307044/4625365

### Steps to reproduce

1. Generate a Capfile and supporting files using `cap install`.
2. Edit the resulting Capfile so that `install_plugin Capistrano::SCM::Git` comes before `require "capistrano/deploy"` (see example below).
3. Try to execute a `cap` command.

```
$ cap production doctor
(Backtrace restricted to imported tasks)
cap aborted!
Don't know how to build task 'deploy:new_release_path' (see --tasks)
```

Capfile:

```ruby
require "capistrano/setup"
require "capistrano/scm/git"
install_plugin Capistrano::SCM::Git
require "capistrano/deploy"
Dir.glob("lib/capistrano/tasks/*.rake").each { |r| import r }
```

### Expected behavior

Ideally the Capfile would be forgiving of `require` ordering. If this is not practical, then the `cap` command should at least emit a helpful error message that allows the user to easily identify and fix the problem.

### Analysis

The SCM plugins use `after "deploy:new_release_path", ...` to install their behavior. If the deploy tasks haven't been loaded yet, this fails.

We could perhaps:

* Allow `after` to register hooks even for non-existent tasks (i.e. defer the hook until the task is defined later). Downside: if you make a typo in a task name, you no longer get an error.
* Add `require "capistrano/deploy"` to the top of `git.rb`, since it is technically a dependency. Downside: someone someday may want to use the Git plugin but provide their own deploy tasks instead of the default Capistrano ones; this change would couple them together.
* Make no changes at all; this is user error.
* Catch the exception and somehow print a better error message for this scenario.

**Comment by mattbrictson** [comment]
An additional aspect to this issue: When upgrading an existing project to 3.7.0, the user is given this instruction:

```
[Deprecation Notice] `set :scm, :git` is deprecated.
To ensure your project is compatible with future versions of Capistrano,
remove the :scm setting and instead add these lines to your Capfile:

    require "capistrano/scm/git"
    install_plugin Capistrano::SCM::Git
```

The message does not say *where* in the Capfile these lines should be added, so it is understandable that users could make a mistake.

**Comment by mcblum** [comment]
This messed me up for quite some time until I found this. For whatever reason, in most projects, it was fine where I put it but in one it was not. It would be awesome to add directions to the notification that fires after the upgrade, per @mattbrictson's comment.

**Comment by mattbrictson** [comment]
> It would be awesome to add directions to the notification that fires after the upgrade

I agree! Would you be willing to contribute a small PR to make this tweak?

**Comment by mcblum** [comment]
@mattbrictson I will give me my best shot. I am the absolute worst at Ruby but maybe this is more Bash. Anyway, why not, right?

**Comment by mcblum** [comment]
@mattbrictson tried to make a pull request to at least expand on the error but got "you are not authorized to do this."

**Comment by mattbrictson** [comment]
Hmm, there shouldn't be any restrictions for opening a PR.

Try this: go to the file you want to modify ([scm_resolver.rb](https://github.com/capistrano/capistrano/blob/master/lib/capistrano/configuration/scm_resolver.rb)) and click the pencil icon to launch the web based editor. GitHub should automatically fork the capistrano repository so you can make an edit. Once you save the changes, GitHub should prompt you to open a PR.

**Comment by robd** [comment]
I just ran into this and I found the experience rather annoying. Specifically:

a) That the Git repo integration has been removed from the core necessitating me changing my Capfile
b) That I had to guess where to put the lines in the Capfile
c) That when I did this and put the lines where it made sense to me (ie as close to the `capistrano/setup` line as possible) I got it wrong and the error message was something obscure

FWIW, I think that making people choose the scm they are using seems like a step in the wrong direction. I know there was a lot of chat about this and I wasn't involved, but it feels to me like this is making our lives as maintainers easier (I guess that's why we're making the change?) at the expense of convenience for the users.

The more we add to the Capfile, the more people like me will get it wrong by putting it in the wrong place, making typos, referring to out of date docs saying the wrong thing, version incompatibilities, etc etc.

Personally, I have a different take on what we should do. I would like to see all of the following removed from the Capfile:

```
require 'capistrano/setup'
require 'capistrano/deploy'

require "capistrano/scm/git"
install_plugin Capistrano::SCM::Git

require 'capistrano/rails'
```

This way there is no risk of a user messing this stuff up.

My hunch is that the majority of users are git and rails users, so my opinion is we should have some way of the plugin system working out the following:

1) it's git because there is a .git repo
2) deploy & rails are needed because I installed the `capistrano-rails` gem

**Comment by leehambley** [comment]
Thanks for the feedback @robd. We discussed a lot whether a little configuration was better than some implicit magic, and settled upon clarity through explicit definition vs. magical inference of the should-be settings.

I appreciate the alternative point of view but don't feel compelled to change anything, but I'll let the issue stand to allow the others to weigh-in with their opinions.

**Comment by mattbrictson** [comment]
I'd like to keep the discussion here focused on fixing the issue at hand, which is a hiccup during the upgrade due to the fact that:

1. It is not clear where in the Capfile the lines should be added
2. There isn't a good error message when they are put in the wrong place
3. As much as possible Capfile *shouldn't care* where the lines are placed

I think most if not all of these problems can be easily solved without completely re-architecting the Capfile, so I'd prefer starting with simple solutions first.

I can create a separate issue to discuss making Capistrano more convention- and less configuration-driven.

**Comment by robd** [comment]
Hey - sorry for the delay in getting back about this. @mattbrictson Thanks very much for raising the new issue. 

I have had a crack at a PR (#1882) to fix this, but I have recently rebuilt my machine so I couldn't run the build at this point. I eyeballed the features and I didn't see any for this message so perhaps this could be merged as is? If not I can work on getting vagrant etc set up. Let me know what you think.

Personally I wouldn't bother doing anything about these:

> 2. There isn't a good error message when they are put in the wrong place
> 3. As much as possible Capfile shouldn't care where the lines are placed

It feels like if we're going to detect that they put the lines in the wrong place we might as well just use that logic to put them in the right place. Likewise my gut feel is that the magic required to allow the Capfile lines to be placed in random order would be a similar amount of work / complexity to us detecting what plugins etc are needed and including them automatically.
